File: examples/python/agg-hand/handlers/deal_cards.py
"""Handler for DealCards command."""

import sys
from pathlib import Path
from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

sys.path.insert(0, str(Path(__file__).parent.parent.parent / "angzarr"))
sys.path.insert(0, str(Path(__file__).parent.parent))

from angzarr_client.proto.angzarr import types_pb2 as types
from angzarr_client.errors import CommandRejectedError
from angzarr_client.proto.examples import hand_pb2 as hand
from angzarr_client.proto.examples import types_pb2 as poker_types

from .state import HandState
from .game_rules import get_game_rules


def handle_deal_cards(
    command_book: types.CommandBook,
    command_any: Any,
    state: HandState,
    seq: int,
) -> types.EventBook:
    """Handle DealCards command."""
    if state.exists():
        raise CommandRejectedError("Hand already dealt")

    cmd = hand.DealCards()
    command_any.Unpack(cmd)

    if not cmd.players:
        raise CommandRejectedError("No players in hand")

    if len(cmd.players) < 2:
        raise CommandRejectedError("Need at least 2 players")

    # Get game rules for this variant
    rules = get_game_rules(cmd.game_variant)

    # Deal cards
    player_roots = [p.player_root for p in cmd.players]
    deal_result = rules.deal_hole_cards(
        deck=[],
        players=player_roots,
        seed=cmd.deck_seed if cmd.deck_seed else None,
    )

    # Build player cards
    player_cards = []
    for player_root, cards in deal_result.player_cards.items():
        pc = hand.PlayerHoleCards(player_root=player_root)
        for suit, rank in cards:
            pc.cards.append(poker_types.Card(suit=suit, rank=rank))
        player_cards.append(pc)

    # Build event
    event = hand.CardsDealt(
        table_root=cmd.table_root,
        hand_number=cmd.hand_number,
        game_variant=cmd.game_variant,
        dealer_position=cmd.dealer_position,
        dealt_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )
    event.player_cards.extend(player_cards)
    event.players.extend(cmd.players)

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.poker/")

    return types.EventBook(
        cover=command_book.cover,
        pages=[
            types.EventPage(
                num=seq,
                event=event_any,
                created_at=Timestamp(
                    seconds=int(datetime.now(timezone.utc).timestamp())
                ),
            )
        ],
    )

File: examples/rust/agg-hand/src/handlers/deal_cards.rs
//! DealCards command handler.

use rand::prelude::*;
use sha2::{Digest, Sha256};

use angzarr_client::proto::examples::{
    Card, CardsDealt, DealCards, GameVariant, PlayerHoleCards, Rank, Suit,
};
use angzarr_client::proto::{event_page, CommandBook, EventBook, EventPage};
use angzarr_client::{pack_event, CommandRejectedError, CommandResult, UnpackAny};
use prost_types::Any;

use crate::game_rules::get_rules;
use crate::state::HandState;

pub fn handle_deal_cards(
    command_book: &CommandBook,
    command_any: &Any,
    state: &HandState,
    seq: u32,
) -> CommandResult<EventBook> {
    if state.exists() {
        return Err(CommandRejectedError::new("Hand already dealt"));
    }

    let cmd: DealCards = command_any
        .unpack()
        .map_err(|e| CommandRejectedError::new(format!("Failed to decode command: {}", e)))?;

    if cmd.players.len() < 2 {
        return Err(CommandRejectedError::new("Need at least 2 players"));
    }

    // Create and shuffle deck
    let mut deck = create_deck();
    let seed = if cmd.deck_seed.is_empty() {
        let mut rng = rand::thread_rng();
        let mut seed = [0u8; 32];
        rng.fill(&mut seed);
        seed.to_vec()
    } else {
        cmd.deck_seed.clone()
    };
    shuffle_deck(&mut deck, &seed);

    // Get game rules for variant
    let variant = GameVariant::try_from(cmd.game_variant).unwrap_or(GameVariant::TexasHoldem);
    let rules = get_rules(variant);
    let cards_per_player = rules.hole_card_count();

    let mut player_cards = Vec::new();
    let total_cards_to_deal = cmd.players.len() * cards_per_player;

    // Deal hole cards
    for (i, player) in cmd.players.iter().enumerate() {
        let start = i * cards_per_player;
        let end = start + cards_per_player;
        let cards: Vec<Card> = deck[start..end].to_vec();
        player_cards.push(PlayerHoleCards {
            player_root: player.player_root.clone(),
            cards,
        });
    }

    // Remaining deck for community cards / draws
    let remaining_deck: Vec<Card> = deck[total_cards_to_deal..].to_vec();

    let event = CardsDealt {
        table_root: cmd.table_root,
        hand_number: cmd.hand_number,
        game_variant: cmd.game_variant,
        player_cards,
        dealer_position: cmd.dealer_position,
        players: cmd.players,
        dealt_at: Some(angzarr_client::now()),
        remaining_deck,
    };

    let event_any = pack_event(&event, "examples.CardsDealt");

    Ok(EventBook {
        cover: command_book.cover.clone(),
        pages: vec![EventPage {
            sequence: Some(event_page::Sequence::Num(seq)),
            event: Some(event_any),
            created_at: Some(angzarr_client::now()),
        }],
        snapshot: None,
        next_sequence: 0,
    })
}

/// Create a standard 52-card deck.
fn create_deck() -> Vec<Card> {
    let suits = [Suit::Clubs, Suit::Diamonds, Suit::Hearts, Suit::Spades];
    let ranks = [
        Rank::Two,
        Rank::Three,
        Rank::Four,
        Rank::Five,
        Rank::Six,
        Rank::Seven,
        Rank::Eight,
        Rank::Nine,
        Rank::Ten,
        Rank::Jack,
        Rank::Queen,
        Rank::King,
        Rank::Ace,
    ];

    let mut deck = Vec::with_capacity(52);
    for suit in suits {
        for rank in ranks {
            deck.push(Card {
                suit: suit as i32,
                rank: rank as i32,
            });
        }
    }
    deck
}

/// Shuffle the deck using a seed for determinism.
fn shuffle_deck(deck: &mut Vec<Card>, seed: &[u8]) {
    let hash = Sha256::digest(seed);
    let seed_int = u64::from_be_bytes(hash[..8].try_into().unwrap());
    let mut rng = StdRng::seed_from_u64(seed_int);
    deck.shuffle(&mut rng);
}
